search/search.go

```diff
@@ -120,39 +130,68 @@ func Open(name string) (*Index, error) {
 // src must be a non-nil struct pointer or implement the FieldLoadSaver
 // interface.
 func (x *Index) Put(c context.Context, id string, src interface{}) (string, error) {
-	d, err := saveDoc(src)
+	ids, err := x.PutMulti(c, []string{id}, []interface{}{src})
 	if err != nil {
 		return "", err
 	}
-	if id != "" {
-		if !validIndexNameOrDocID(id) {
-			return "", fmt.Errorf("search: invalid ID %q", id)
+	return ids[0], nil
+}
+
+// PutMulti is like Put, but is more efficient for adding multiple documents to
+// the index at once.
+//
+// Up to 200 documents can be added at once. ErrTooManyDocuments is returned if
+// you try to add more.
+//
+// ids can either be an empty slice (which means new IDs will be allocated for
+// each of the documents added) or a slice the same size as srcs.
+func (x *Index) PutMulti(c context.Context, ids []string, srcs []interface{}) ([]string, error) {
+	if len(ids) != 0 && len(srcs) != len(ids) {
+		return nil, fmt.Errorf("search: PutMulti expects ids and srcs slices of the same length")
+	}
+	if len(srcs) > maxDocumentsPerPutDelete {
+		return nil, ErrTooManyDocuments
+	}
+
+	docs := make([]*pb.Document, len(srcs))
+	for i, s := range srcs {
+		var err error
+		docs[i], err = saveDoc(s)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(ids) != 0 && ids[i] != "" {
+			if !validIndexNameOrDocID(ids[i]) {
+				return nil, fmt.Errorf("search: invalid ID %q", ids[i])
+			}
+			docs[i].Id = proto.String(ids[i])
 		}
-		d.Id = proto.String(id)
 	}
+
 	// spec is modified by Call when applying the current Namespace, so copy it to
 	// avoid retaining the namespace beyond the scope of the Call.
 	spec := x.spec
 	req := &pb.IndexDocumentRequest{
 		Params: &pb.IndexDocumentParams{
-			Document:  []*pb.Document{d},
+			Document:  docs,
 			IndexSpec: &spec,
 		},
 	}
 	res := &pb.IndexDocumentResponse{}
 	if err := internal.Call(c, "search", "IndexDocument", req, res); err != nil {
-		return "", err
+		return nil, err
 	}
-	if len(res.Status) > 0 {
-		if s := res.Status[0]; s.GetCode() != pb.SearchServiceError_OK {
-			return "", fmt.Errorf("search: %s: %s", s.GetCode(), s.GetErrorDetail())
+	for _, s := range res.Status {
+		if s.GetCode() != pb.SearchServiceError_OK {
+			return nil, fmt.Errorf("search: %s: %s", s.GetCode(), s.GetErrorDetail())
```

Comment:
Same question here as below -- since this is now a multi put, should we aggregate error messages? Or otherwise indicate that some puts have succeeded while others have not?